### ISSUE TITLE
Label-Placement outside geometry results in wrong hierarchy

### ISSUE THREAD

**Issue by CeallachKN** [whosonfirst-data/whosonfirst-data#2159]
While checking some issues with getting #2032 ready, i've found that if the Label is placed outside the geometry the result is a wrong hierarchy for the record.

Just a few examples:

- [101905935](https://spelunker.whosonfirst.org/id/101905935/) -> localadmin should be [1377688863](https://spelunker.whosonfirst.org/id/1377688863/) but is [1377690117](https://spelunker.whosonfirst.org/id/1377690117)
- [1393702421](https://spelunker.whosonfirst.org/id/1393702421) -> Localadmin should be [1377689731](https://spelunker.whosonfirst.org/id/1377689731) but is [1377689497](https://spelunker.whosonfirst.org/id/1377689497)
- [101826831](https://spelunker.whosonfirst.org/id/101826831) -> Localadmin should be [1377689757](https://spelunker.whosonfirst.org/id/1377689757) but is [1377689613](https://spelunker.whosonfirst.org/id/1377689613)
- [101910073](https://spelunker.whosonfirst.org/id/101910073) -> localadmin should be [1377688845](https://spelunker.whosonfirst.org/id/1377688845) but is [1377689145](https://spelunker.whosonfirst.org/id/1377689145)

My assumption is that the label-centroid is used (because if a geometry has an odd shape the math-centroid won't work) to determine the hierarchy. For Pelias this results in an issue described in https://github.com/whosonfirst-data/whosonfirst-data/issues/2036#issuecomment-1601058246 
@missinglink : this should be of interest for you as well.

I haven't checked other areas besides Bavaria-records in https://github.com/whosonfirst-data/whosonfirst-data-admin-de  but if my assumption is correct, that could be a potential issue for all wof-data repositories.

I can expand the list of examples if needed, since a check if the label-centroid is inside the records own geometry should be pretty simple but unfortunately I'm a bit short on time right now to do this on a larger scale. 
Since the label-placement is manual afaik, i suppose this would result in a lot of manual work as well.

**Comment by missinglink** [comment]
Hmmm yeah, this is clearly a bug.
I believe [mapshaper](https://github.com/mbloch/mapshaper) is being used to generate these label centroids.

<img width="543" alt="Screenshot 2023-09-06 at 16 04 46" src="https://github.com/whosonfirst-data/whosonfirst-data/assets/738069/e3872216-2e6d-4845-97fb-99d10fbd13e6">

**Comment by missinglink** [comment]
I had a deeper look into this and unfortunately it seems to affect many records, steps to reproduce:

```sql
SELECT
  id,
  json_extract(body, '$.properties.lbl:latitude') AS lat,
  json_extract(body, '$.properties.lbl:longitude') AS lon
FROM geojson
WHERE lat IS NOT NULL AND lon IS NOT NULL
AND json_extract(body, '$.properties.mz:is_current') = 1
AND json_extract(body, '$.geometry.type') != 'Point'
AND Contains(
  GeomFromGeoJSON(json_extract(body, '$.geometry')),
  MakePoint(lon, lat)
) != 1;
```

```bash
spatialite -silent whosonfirst-data-admin-latest.db < label-check.sql
```

example of some records where the label centroid falls outside the geometry:

```
85764185|-34.591462|-58.532595
85764259|-34.553289|-58.493739
85765603|-34.670624|-58.388802
102049871|48.192288|16.096632
85898351|48.246766|16.311108
85898393|48.224223|16.445321
85898693|48.233835|16.417596
85898697|48.195221|16.393025
85898699|48.21012|16.411951
85898733|48.184102|16.298894
101929683|-33.476508|150.123596
101929969|-34.439055|150.887707
101931319|-34.404984|150.887558
101931329|-34.446247|150.869506
101931623|-34.481401|150.909402
101931779|-34.364585|150.896869
101931803|-32.861162|151.708158
101931817|-34.269703|150.959015
101931967|-32.902024|151.779011
101932127|-34.389075|150.907359
101932341|-32.916821|151.759992
101932505|-37.781407|144.912798
...
```

**Comment by nvkelso** [comment]
Thanks for the report @CeallachKN and for the repo SQL @missinglink. I agree this is a problem WOF should address.

My guess at some point in the past we updated the default geometry but didn't update the mapshaper label centroid, and we prefer the mapshaper centroid over the math centroid because it _should_ be in the feature (where the math centroid is sometimes out of the centroid for irregular shapes).

@stepps00 can you use the list of WOF IDs that @missinglink suggests (using his SQL) to regenerate MapShaper centroids and then re-PIP those places?

**Comment by missinglink** [comment]
I can save you some time:
https://gist.github.com/missinglink/0662d6b500055614eb5349c4638b5a88

**Comment by nvkelso** [comment]
Thanks! So only 4k of them :)

**Comment by stepps00** [comment]
Thanks for the list of affected records, @missinglink 

The set of PRs linked above should solve most of these issues using new Mapshaper-sourced inner-centroids. I'll take a bit of time to QA some of these changes and merge if things look good.

I encountered a handful of errors, example: https://spelunker.whosonfirst.org/id/102005153/

Note there will be some number of records like this that will still fail the `Contains` check once the PRs are merged.

**Comment by missinglink** [comment]
We could change the spatial predicate from `Contains` to `Intersects` and it should work for all geometry types including Points & Lines.

**Comment by nvkelso** [comment]
@missinglink can you link to PR that changes the spatial predicate from `Contains` to `Intersects`?

@stepps00 gentle nudge on PRs which might get stale / conflict with my incoming PRs (which should not affect geometry, but as these things go)...

**Comment by missinglink** [comment]
I'm really not familiar with the old python code which I believe is still be used for the PIP work.

It should be easy enough for someone who knows it, or I can make an attempt at it but I don't have the knowledge to ensure it's tested and working, I never did any professional python coding.

If @stepps00 is using the Go code then it's much more familiar for me.

**Comment by thisisaaronland** [comment]
I am familiar with both but I am not what the context for "spatial predicates" is. I assume we're talking about the Mapshaper logic?

Can someone elaborate?

**Comment by missinglink** [comment]
We just need some way of making sure the label centroids are actually inside the shape because they are being used later to establish the admin hierarchy.

**Comment by thisisaaronland** [comment]
Right, I understand that. The label centroids are generated using Mapshaper's "points inner" command, for example:

https://github.com/whosonfirst/py-mapzen-whosonfirst-mapshaper/blob/master/mapzen/whosonfirst/mapshaper/__init__.py#L100-L117

It's entirely possible that there are bugs or edge cases in Matthew Bloch's code but, at least until now, it's been a kind of matter of faith that his code will do a better job than anything/one else at deriving a suitable label centroid.
 
What I don't understand is what "spatial predicates" have to do with anything since they aren't specifically relevant to the Mapshaper command in questin.

**Comment by missinglink** [comment]
> What I don't understand is what "spatial predicates" have to do with anything since they aren't specifically relevant to the Mapshaper command in questin.

Sorry that was unclear, I meant the SQL query I wrote above could be modified such that changing the spatial predicate would allow removing of the `json_extract(body, '$.geometry.type') != 'Point'` clause.

> it's been a kind of matter of faith that his code will do a better job than anything/one else at deriving a suitable label centroid.

There surely must be some geospatial function which can guarantee a generated centroid lies inside the geometry, something like [PointOnSurface](https://postgis.net/docs/ST_PointOnSurface.html)?

**Comment by missinglink** [comment]
![Screenshot_20230927-231818](https://github.com/whosonfirst-data/whosonfirst-data/assets/738069/2812e2a0-65d4-44a5-b77e-3f623d6709bb)

**Comment by missinglink** [comment]
I think it's actually ok that the label centroid falls outside the shape, this might actually be better for display in many cases.

The issue is more for the reverse geo centroid, which is problematic in cases such as where the polygon is conterminous to its parent but the centroid doesn't intersect it.

**Comment by thisisaaronland** [comment]
The `reversegeo` centroid is not (should not) be derived from Mapshaper. That is derived using traditional point-in-polygon / raycasting techniques.

**Comment by nvkelso** [comment]
We're still interested in fixing this. The PRs have gone stale and have conflicts now. We plan to remake the PRs in July. There will be some complication with PIPing those PRs we'll need to sort out.
